### PR TITLE
Introduce InsightsClient

### PR DESCRIPTION
**Pull request by julienbourdeau**
| Q                 | A
| ----------------- | ----------
| Bug fix?          | no
| New feature?      | yes    
| BC breaks?        | no     
| Need Doc update   | yes

## Describe your change

All Algolia clients are going to allow access to all our rest API. This PR adds support for Insights API.

### How to use

The insights API has 1 endpoint `POST /1/events` which can receive a few different events. We discussed the DX with the InstantSearch team (mobile and web) and we thought the api clients should make it easier to prepare your payload for each events (click, conversion, and view).

In the end we'd like a DX like

```php
$insights = \Algolia\AlgoliaSearch\InsightsClient::create($appId, $apiKey, 'us');

// UserToken ca be set so all following operations will use it
$insights->setUserToken('30d69d863dde8');

$insights->search($queryId)->click([
    'eventName' => SomeClass::SOME_CONST,
    'indexName' => 'products',
    'objectID' => 'someProductObjectID',
    'position' => 2,
]);

$insights->visit()->view([
    'eventName' => 'open single product page',
    'indexName' => 'products',
    'objectID' => 'someProductObjectID',
]);


$insights->search($queryId)->conversion([
    'eventName' => 'Added to cart',
    'indexName' => 'products',
    'objectID' => 'someProductObjectID',
]);
```

We split events between `visit` and `search` because for search must have a queryId (maybe the implem' makes it a bit clearer).

Another way ot use the client is by building the event yourself. This method would typically be used for batching.

```php
$insights = \Algolia\AlgoliaSearch\InsightsClient::create($appId, $apiKey, 'us');

$insights->addEvent(
    'userToken' => '30d69d863dde8',
    'queryID' => $queryId,
    'eventType' => 'conversion'
    'eventName' => 'Added to cart',
    'indexName' => 'products',
    'objectIDs' => ['someProductObjectID'],
    'positions' => [2],
    'timestamp' => time(),
]);
```

Note that `objectIDs` and `positions` must be arrays only our helper methods accept `objectID` and `position` in a singular form.

If you do the following, what's passed in the $event will override what was set beofre.

```php
$insights = \Algolia\AlgoliaSearch\InsightsClient::create($appId, $apiKey, 'us');

// UserToken ca be set so all following operations will use it
$insights->setUserToken('30d69d863dde8');

$insights->search($queryId)->click([
    'userToken' => '30d69d863dde8',
    'queryID' => $anotherQueryID,
    'eventName' => 'Added to cart',
    'indexName' => 'products',
    'objectIDs' => ['someProductObjectID'],
    'positions' => [2],
]);
```

### Future-proof

There will always be this method available if you want to define the events yourself:

```php
$insights->sendEvent($fullEvents);
$insights->sendEvents($arrayOfEvents);
```

### CODE REVIEW

**Comment by julienbourdeau** [comment]
After discussing with @maxiloc yesterday, we made some modifications. In the first proposal, we decided to hide what was an event for analytics and what events were for Personalization. We challanged this idea, it might not be the best way.

Also, we should be moving to method arguments rather than 1 hashmap.

We came up with something like this:

```php
//Analytics
$insights->searchAnalytics($queryId)->click($indexName, $objectID, $position, $userToken);
$insights->searchAnalytics($queryId)->conversion($indexName, $objectID, $position, $userToken);
```

```php
$insights->personalizeForUser($userToken)->clicked($eventName, $indexName, $objectIDorFilters, $requestOptions);
$insights->personalizeForUser($userToken)->viewed($eventName, $indexName, $objectIDorFilters, $requestOptions);
$insights->personalizeForUser($userToken)->converted($eventName, $indexName, $objectIDorFilters, $requestOptions);
```

Questions left to answer

* Can you send one event for both analytics and personalization?
* Is `$userToken` really required for analytics?
* Can we have a default `eventName` for analytics?

Feedback is more than welcome.

File: src/Insights/PersoInsightClient.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
+<?php
+
+namespace Algolia\AlgoliaSearch\Insights;
+
+final class PersoInsightClient extends AbstractInsightsClient
```
Comment (rayrutjes) [inline src/Insights/PersoInsightClient.php]: No `s` here?

File: src/Insights/AbstractInsightsClient.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+        $payload = array('events' => $events);
+
+        return $this->api->write('POST', Helpers::apiPath('/1/events'), $payload, $requestOptions);
+    }
+
+    private function reformatEvent($e)
+    {
+        if (!isset($e['timestamp'])) {
+            $e['timestamp'] = time();
```
Comment (rayrutjes) [inline src/Insights/AbstractInsightsClient.php]: I think this needs to be in milliseconds.

File: src/Insights/UserInsightsClient.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            'eventName' => $eventName,
+            'index' => $indexName,
+        ));
+
+        return $this->sendEvent($event, $requestOptions);
+    }
+
+    private function sendEvent($event, $requestOptions = array())
+    {
+        if (isset($requestOptions['timestamp'])) {
```
Comment (julienbourdeau) [inline src/Insights/UserInsightsClient.php]: should be removed

File: src/Insights/UserInsightsClient.php [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        $event = array(
+            'objectIDs' => $objectIDs,
+            'queryID' => $queryID,
+            'positions' => $positions,
+        );
+
+        return $this->clicked($event, $eventName, $indexName, $requestOptions);
+    }
+
+    private function clicked($event, $eventName, $indexName, $requestOptions)
```
Comment (nunomaduro) [inline src/Insights/UserInsightsClient.php]: Nice refactor! I will take this in consideration for the Python V2.

File: src/Insights/UserInsightsClient.php [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        $event = array_merge($event, array(
+            'type' => 'click',
+            'eventName' => $eventName,
+            'index' => $indexName,
+        ));
+
+        return $this->sendEvent($event, $requestOptions);
+    }
+
+    public function convertedObjectIDs($eventName, $indexName, $objectIDs, $requestOptions = array())
```
Comment (nunomaduro) [inline src/Insights/UserInsightsClient.php]: I think you may have forgotten `convertedFilters`.

Comment (julienbourdeau) [reply src/Insights/UserInsightsClient.php]: oops 😓

File: src/InsightsClient.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    public function sendEvent($event, $requestOptions = array())
+    {
+        return $this->sendEvents(array($event), $requestOptions);
+    }
+
+    public function sendEvents($events, $requestOptions = array())
+    {
+        $payload = array('events' => $events);
+
+        return $this->api->write('POST', api_path('/1/events'), $payload, $requestOptions);
```
[the remark this answers is not in the document]
Comment (nunomaduro) [reply src/InsightsClient.php]: I believe we should remove the `return` here.

File: src/InsightsClient.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public function user($userToken)
+    {
+        return new UserInsightsClient($this, $userToken);
+    }
+
+    public function sendEvent($event, $requestOptions = array())
+    {
+        return $this->sendEvents(array($event), $requestOptions);
```
Comment (nunomaduro) [inline src/InsightsClient.php]: I believe we should remove the `return` here.

File: src/InsightsClient.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (julienbourdeau) [reply src/InsightsClient.php]: I think response objects should only be used if there is a need for `waitable`. I this case dont think there is even a response body.

File: src/InsightsClient.php [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        $this->api = $api;
+        $this->config = $config;
+    }
+
+    public static function create($appId = null, $apiKey = null, $region = null, $userToken = null)
+    {
+        $config = InsightsConfig::create($appId, $apiKey, $region);
+
+        if ($userToken) {
+            $config->setUserToken($config);
```
[the remark this answers is not in the document]
Comment (julienbourdeau) [reply src/InsightsClient.php]: acutally it should be removed ^^

File: src/Insights/UserInsightsClient.php [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    {
+        return $this->clicked(array('filters' => $filters), $eventName, $indexName, $requestOptions);
+    }
+
+    public function clickedObjectIDs($eventName, $indexName, $objectIDs, $requestOptions = array())
+    {
+        return $this->clicked(array('objectIDs' => $objectIDs), $eventName, $indexName, $requestOptions);
+    }
+
+    public function clickedObjectIDsAfterSearch($eventName, $indexName, $objectIDs, $queryID, $positions, $requestOptions = array())
```
Comment (nunomaduro) [inline src/Insights/UserInsightsClient.php]: The signature should be: `($eventName, $indexName, $objectIDs, $positions, $queryID, $requestOptions = array())

File: src/Config/InsightsConfig.php [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+        return $this;
+    }
+
+    public function getRegion()
+    {
+        return $this->config['region'];
+    }
+
+    public function setUserToken($userToken)
```
Comment (nunomaduro) [inline src/Config/InsightsConfig.php]: This code is used?

File: src/Config/InsightsConfig.php [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        $this->config['userToken'] = $userToken;
+
+        return $this;
+    }
+
+    public function getUserToken()
```
Comment (nunomaduro) [inline src/Config/InsightsConfig.php]: This code is used?

**Pull request merged**